### PR TITLE
Add On-Stack Replacement for Sulong.

### PR DESCRIPTION
**Pull request by rmosaner**
This PR adds On-Stack Replacement (OSR) for Sulong.

This PR was transferred from [https://github.com/graalvm/sulong/pull/859](url)

### CODE REVIEW

File: sulong/tests/com.oracle.truffle.llvm.tests.debug/debug/testLoop.c [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+ * GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+ * NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
+ * OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+__attribute__((constructor)) int start()
+{
+    int j = 0;
+    int i = 0;
+    for(i = 0; i < 5; i++) {
```
Comment (jkreindl) [inline sulong/tests/com.oracle.truffle.llvm.tests.debug/debug/testLoop.c]: There are some spaces between braces and keywords missing.

File: sulong/tests/com.oracle.truffle.llvm.tests.debug/trace/testLoop.txt [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+        MEMBER int "int" "i" "4"
+        MEMBER int "int" "k" "0"
+
+STOP 61 CONTINUE "start"
+    OPEN_SCOPE
+	MEMBER int "int" "j" "10"
+        MEMBER int "int" "i" "5"
+        MEMBER int "int" "k" "0"
+
+BREAK 65 CONTINUE "start"
```
Comment (jkreindl) [inline sulong/tests/com.oracle.truffle.llvm.tests.debug/trace/testLoop.txt]: The C file for this test only has 61 lines, you cannot stop at line 65. Did you perhaps reformat the C code without changing the trace?

**Comment by fniephaus** [comment]
What's the progress on this PR? Why hasn't this been merged yet? Also, how (if at all) is the new [BlockNode](https://www.graalvm.org/truffle/javadoc/com/oracle/truffle/api/nodes/BlockNode.html) related to this?

**Comment by rschatz** [comment]
Merged in https://github.com/oracle/graal/commit/a648506e90d1deeb6d6bfa023896a092c29997ba

**Comment by fniephaus** [comment]
Fantastic!
@rschatz and @lukasstadler: IIRC this PR was delayed because the approach [decreased peak performance](https://arxiv.org/pdf/1909.08815.pdf). Has this been addressed and if yes, how and where?

**Comment by rschatz** [comment]
It might still decrease peak performance a little bit, but the worst issues seem to have disappeared. We didn't investigate too deeply why, but there have been several compiler improvements in the meantime that might have improved it.

**Comment by fniephaus** [comment]
I remember that @davleopo had some ideas how to work around this in the compiler. Maybe he knows more about it now?

**Pull request closed without merging**